File: app/src/main/java/io/plan8/backoffice/vm/MoreFragmentVM.java
package io.plan8.backoffice.vm;

import android.databinding.Bindable;
import android.databinding.ViewDataBinding;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.InputType;
import android.view.View;

import com.afollestad.materialdialogs.DialogAction;
import com.afollestad.materialdialogs.MaterialDialog;

import java.util.ArrayList;
import java.util.List;

import io.plan8.backoffice.ApplicationManager;
import io.plan8.backoffice.BR;
import io.plan8.backoffice.R;
import io.plan8.backoffice.adapter.BindingRecyclerViewAdapter;
import io.plan8.backoffice.fragment.BaseFragment;
import io.plan8.backoffice.model.BaseModel;
import io.plan8.backoffice.model.api.Member;
import io.plan8.backoffice.model.api.User;
import io.plan8.backoffice.model.item.LabelItem;
import io.plan8.backoffice.vm.item.EmptySpaceItemVM;
import io.plan8.backoffice.vm.item.LabelItemVM;
import io.plan8.backoffice.vm.item.MoreProfileItemVM;
import io.plan8.backoffice.vm.item.MoreTeamItemVM;

/**
 * Created by SSozi on 2017. 11. 28..
 */

public class MoreFragmentVM extends FragmentVM {
    private BindingRecyclerViewAdapter<BaseModel> adapter;
    private List<BaseModel> moreItemList = new ArrayList<>();
    private boolean completedLoading;
    private boolean swipeFlag = true;

    public MoreFragmentVM(BaseFragment fragment, @Nullable final Bundle savedInstanceState) {
        super(fragment, savedInstanceState);

        adapter = new BindingRecyclerViewAdapter<BaseModel>() {
            @Override
            protected int selectViewLayoutType(BaseModel data) {
                if (data instanceof LabelItem) {
                    return R.layout.item_more_title;
                } else if (data instanceof User) {
                    return R.layout.item_more_profile;
                } else if (data instanceof Member) {
                    return R.layout.item_more_team;
                } else {
                    return R.layout.item_empty_space;
                }
            }

            @Override
            protected void bindVariables(ViewDataBinding binding, BaseModel data) {
                if (data instanceof LabelItem) {
                    binding.setVariable(BR.vm, new LabelItemVM(getFragment(), savedInstanceState, (LabelItem) data));
                } else if (data instanceof User) {
                    binding.setVariable(BR.vm, new MoreProfileItemVM(getFragment(), savedInstanceState, (User) data));
                } else if (data instanceof Member) {
                    binding.setVariable(BR.vm, new MoreTeamItemVM(getFragment(), savedInstanceState, (Member) data));
                } else {
                    binding.setVariable(BR.vm, new EmptySpaceItemVM(getFragment(), savedInstanceState));
                }
            }
        };

        setData(moreItemList);
    }

    public RecyclerView.LayoutManager getLayoutManager() {
        return new LinearLayoutManager(getFragment().getContext(), LinearLayoutManager.VERTICAL, false);
    }

    public RecyclerView.Adapter getAdapter() {
        return adapter;
    }

    public void setData(List<BaseModel> data) {
        if (null != adapter) adapter.setData(data);
    }

    public void logout(final View view) {
        new MaterialDialog.Builder(getFragment().getContext())
                .title("정말로 로그아웃을 하시겠습니까?")
                .inputType(InputType.TYPE_CLASS_TEXT)
                .positiveText("확인")
                .negativeText("취소")
                .onPositive(new MaterialDialog.SingleButtonCallback() {
                    @Override
                    public void onClick(@NonNull MaterialDialog dialog, @NonNull DialogAction which) {
                        ApplicationManager.getInstance().logout();
                        dialog.dismiss();
                    }
                })
                .show();
    }

    @Bindable
    public boolean isCompletedLoading() {
        return completedLoading;
    }

    public void setCompletedLoading(boolean compltedLoading) {
        this.completedLoading = compltedLoading;
        notifyPropertyChanged(BR.completedLoading);
    }

    @Bindable
    public boolean getSwipeFlag() {
        return swipeFlag;
    }

    public void setSwipeFlag(boolean flag) {
        swipeFlag = flag;
        notifyPropertyChanged(BR.swipeFlag);
    }
}

File: app/src/main/java/io/plan8/backoffice/vm/MoreFragmentVM.kt
package io.plan8.backoffice.vm

import android.content.Intent
import android.databinding.ViewDataBinding
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View
import android.widget.Toast
import io.plan8.backoffice.BR
import io.plan8.backoffice.R
import io.plan8.backoffice.SharedPreferenceManager
import io.plan8.backoffice.activity.LoginActivity
import io.plan8.backoffice.adapter.BindingRecyclerViewAdapter
import io.plan8.backoffice.model.item.TaskItem
import io.plan8.backoffice.model.item.MoreProfileItem
import io.plan8.backoffice.model.item.MoreTeamItem
import io.plan8.backoffice.model.item.MoreTitleItem
import io.plan8.backoffice.vm.item.*

/**
 * Created by chokwanghwan on 2017. 11. 9..
 */
open class MoreFragmentVM(val fragment: Fragment
                          , savedInstanceState: Bundle?, moreItemList: List<Any>) : FragmentVM(fragment, savedInstanceState) {
    var adapter: BindingRecyclerViewAdapter<Any>
    var data: List<Any>? = null
    lateinit var moreProfileVM: MoreProfileItemVM

    init {
        adapter = object : BindingRecyclerViewAdapter<Any>() {

            override fun selectViewLayoutType(data: Any): Int {
                return when (data) {
                    is MoreTitleItem -> R.layout.item_more_title
                    is MoreProfileItem -> R.layout.item_more_profile
                    is MoreTeamItem -> R.layout.item_more_team
                    else -> R.layout.item_empty_space
                }
            }

            override fun bindVariables(binding: ViewDataBinding, data: Any) {
                when (data) {
                    is MoreTitleItem -> {
                        val moreTitleItemVM = MoreTitleItemVM(fragment, savedInstanceState, data)
                        binding.setVariable(BR.vm, moreTitleItemVM)
                    }
                    is MoreProfileItem -> {
                        val moreProfileItemVM = MoreProfileItemVM(fragment, savedInstanceState, data)
                        binding.setVariable(BR.vm, moreProfileItemVM)
                        moreProfileVM = moreProfileItemVM
                    }
                    is MoreTeamItem -> {
                        val moreTeamItemVM = MoreTeamItemVM(fragment, savedInstanceState, data)
                        binding.setVariable(BR.vm, moreTeamItemVM)
                    }
                    else -> {
                        val emptySpaceItemVM = EmptySpaceItemVM(fragment, savedInstanceState)
                        binding.setVariable(BR.vm, emptySpaceItemVM)
                    }
                }
            }
        }

        setDatas(moreItemList)
    }

    fun getLayoutManager(): RecyclerView.LayoutManager {
        return LinearLayoutManager(context)
    }

    fun setDatas(data: List<Any>?) {
        this.data = data
        adapter.data = this.data
    }

    fun logout(view: View){
        SharedPreferenceManager(fragment.context).removeToken()
        val loginIntent = Intent(fragment.activity, LoginActivity::class.java)
        fragment.activity.startActivity(loginIntent)
        fragment.activity.finish()
        fragment.activity.overridePendingTransition(R.anim.pull_in_left_activity, R.anim.push_out_right_activity)
    }
}